### DISCUSSION TITLE
Using URLRouting in a Swift 6 environment

### DISCUSSION THREAD

**Discussion by dirk68-work** [pointfreeco/swift-url-routing#110] [General]
Hi,

I tried to use URLRouting in a Swift 6 environment and soon get concurrency errors.
For example: The Router types seem not to be sendable. If you try to define one you get the following error:

```swift
let appRouter = OneOf {
        // GET /books
        Route(.case(AppRoute.books)) {
            Path { "books" }
        }
        
        // GET /books/:id
        Route(.case(AppRoute.book(id:))) {
            Path { "books"; Digits() }
        }
    }
```

> Stored property 'appRouter' of 'Sendable'-conforming struct 'InventoryResponder' has non-sendable type 'OneOf<URLRequestData, AppRoute, OneOfBuilder<URLRequestData, AppRoute>.OneOf2<Route<Parsers.MapConversion<Path<PathBuilder.Component<String>>, AnyCasePath<AppRoute, ()>>>, Route<Parsers.MapConversion<Path<PathBuilder.PathZipVO<String, Digits<PartialRangeFrom<Int>, Conversions.SubstringToUTF8View>>>, AnyCasePath<AppRoute, Int>>>>>'

What is the right way do do this or is the library only usable with
`@preconcurrency import URLRouting` ?

**Comment by dirk68-work** [comment]
Does anybody have any advice on that?

**Reply by lukeredpath** [reply]
Currently the only solution is to use `@preconcurrency import` if you need to pass a router across a boundary e.g. into a sendable closure. The long-term fix is for the library to mark its types as Sendable, which most if not all should be fairly trivially.

**Comment by junebash** [comment]
You could also 

- use `nonisolated(unsafe) let router = ...`
- use a computed property (eg `var router: some URLRouter<MyRoute> { ... }`)
